files/xml.py

```diff
@@ -0,0 +1,470 @@
+    required: false
+    default: false
+    choices: []
+  content:
+    description:
+      - "Search for a given C(xpath) and get content"
+    required: false
+    default: false
+    choices:
+      - "attribute"
+      - "text"
+requirements:
+    - The remote end must have the Python C(lxml) library installed
+author: Tim Bielawa, Magnus Hedemark
+'''
+
+
+from io import BytesIO
+from lxml import etree
+try:
+    import json
+except:
+    import simplejson as json
+import lxml
+import os
+
+def print_match(module, tree, xpath, namespaces):
+    match = tree.xpath(xpath, namespaces=namespaces)
+    match_xpaths = []
+    for m in match:
+        match_xpaths.append(tree.getpath(m))
+    match_str = json.dumps(match_xpaths)
+    msg = "selector '%s' match: %s" % (xpath, match_str)
+    finish(module, tree, xpath, namespaces, changed=False, msg=msg)
+
+def count(module, tree, xpath, namespaces):
+    """ Return the count of nodes matching the xpath """
+    hits = tree.xpath("count(/%s)" % xpath, namespaces=namespaces)
+    finish(module, tree, xpath, namespaces, changed=False, msg=int(hits), hitcount=int(hits))
+
+def is_node(tree, xpath, namespaces):
+    """ Test if a given xpath matches anything and if that match is a node.
+
+    For now we just assume you're only searching for one specific thing."""
+    if xpath_matches(tree, xpath, namespaces):
+        # OK, it found something
+        match = tree.xpath(xpath, namespaces=namespaces)
+        if type(match[0]) == lxml.etree._Element:
+            return True
+
+    return False
+
+def is_attribute(tree, xpath, namespaces):
+    """ Test if a given xpath matches and that match is an attribute
+
+An xpath attribute search will only match one item"""
+    if xpath_matches(tree, xpath, namespaces):
+        match = tree.xpath(xpath, namespaces=namespaces)
+        if type(match[0]) == lxml.etree._ElementStringResult:
+            return True
+    return False
+
+def xpath_matches(tree, xpath, namespaces):
+    """ Test if a node exists """
+    if tree.xpath(xpath, namespaces=namespaces):
+        return True
+    else:
+        return False
+
+def delete_xpath_target(module, tree, xpath, namespaces):
+    """ Delete an attribute or element from a tree """
+    try:
+        for result in tree.xpath(xpath, namespaces=namespaces):
+            if not module.check_mode:
+                # Get the xpath for this result
+                if is_attribute(tree, xpath, namespaces):
+                    # Delete an attribute
+                    parent = result.getparent()
+                    # Pop this attribute match out of the parent
+                    # node's 'attrib' dict by using this match's
+                    # 'attrname' attribute for the key
+                    parent.attrib.pop(result.attrname)
+                elif is_node(tree, xpath, namespaces):
+                    # Delete an element
+                    result.getparent().remove(result)
+    except Exception, e:
```

Comment:
making modules compatible with python3 is something that we're slowly working towards.  Part of that effort requires that exceptions be valid python 3 syntax.  This can be troublesome when a module also has to be compatible with pyhton-2.4.  So we have a helper function that can be used:
```
[[Somewhere at the top, with your imports]]
from ansible.module_utils.pycompat24 import get_exception

[[ Here, where you're catching the exception:]]
except Exception:
    e = get_excetion()
```